blog/en/blog/2022/08/17/backend-for-frontend-demo.md

```diff
@@ -0,0 +1,299 @@
+---
+title: "Backend-for-Frontend: the demo"
+authors:
+  - name: Nicolas Fränkel
+    title: Author
+    url: https://github.com/nfrankel
+    image_url: https://avatars.githubusercontent.com/u/752258
+keywords: 
+- API gateway
+- Apache APISIX
+- System architecture
+- Mciroservices
+description: This article describes a demo code to implement the Backend-For-Frontend pattern.
+tags: [Case Studies]
+---
+
+> This article describes a demo code to implement the Backend-For-Frontend pattern.
+
+<!--truncate-->
+
+<head>
+    <link rel="canonical" href="https://blog.frankel.ch/backend-for-frontend-demo/" />
+</head>
+
+In [one of my earlier posts](https://blog.frankel.ch/backend-for-frontend/), I described the Backend-for-Frontend pattern. In short, it offers a single facade over multiple backend parts. Moreover, it provides each client type, _e.g._ desktop, mobile, exactly the data that it needs and not more in the format required by this client type.
+
+## The use-case
+
+Imagine the following use-case. In a e-commerce shop, the home page should display multiple *unrelated* data at once.
+
+* Products: The business could configure which items are shown on the home page. They could be generic, "hero" products, or personalized, products that the customer ordered previously.
+* News: Again, the newsfeed could be generic or personalized.
+* Profile-related information
+* Cart content
+* Non-business related information, such as build number, build timestamp, version, etc.
+
+Depending on the client, we want more or less data. For example, on a client with limited display size, we probably want to limit a product to its name and its image. On the other hand, on desktop, we are happy to display both the above, plus a catch phrase (or a more catchy - and longer - name) and the full description.
+
+Every client requires its specific data and for performance reasons, we want to fetch them in a single call. It sounds like a use-case for <abbr title="Backend-for-Frontend">BFF</abbr>.
+
+## Setting up the demo
+
+In order to simplify things, I'll keep only three sources of data: products, news and technical data. Three unrelated data sources are enough to highlight the issue.
+
+In the demo, I'm using Python and Flask, but the underlying technology is irrelevant, since BFF is an architectural pattern.
+
+The initial situation is a monolith. The monolith offers an endpoint for each data source, and a single aggregating endpoint for all of them:
+
+```python
+@app.route("/")
+def home():
+    return {
+      'products': products,       #1
+      'news': news,               #1
+      'info': debug               #1
+  }
+```
+
+1. Somehow get the data internally, _e.g._, from the database
```

Comment:
```suggestion
Somehow get the data internally, _e.g._, from the database.
```

Reply:
Because there is no follow-up second step, I think it will be a bit awkward to put a 1 here, it is better to change it directly to the text format description
<img width="870" alt="image" src="https://user-images.githubusercontent.com/39793568/185270856-cec76318-2d8d-4f41-a252-f8d9168a04f5.png">